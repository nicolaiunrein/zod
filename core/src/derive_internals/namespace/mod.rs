use darling::FromDeriveInput;
use proc_macro2::TokenStream;
use quote::quote;
use syn::DeriveInput;

use crate::utils::zod_core;

#[derive(FromDeriveInput)]
#[darling(
    attributes(zod),
    forward_attrs(allow, doc, cfg),
    supports(struct_named, struct_newtype, struct_tuple, struct_unit)
)]
struct NamespaceAttrs {
    ident: syn::Ident,
    name: Option<String>,
}

pub fn expand(input: TokenStream) -> TokenStream {
    let derive_input: DeriveInput = match syn::parse2(input) {
        Ok(input) => input,
        Err(err) => return err.into_compile_error(),
    };
    let NamespaceAttrs { name, ident } = match NamespaceAttrs::from_derive_input(&derive_input) {
        Ok(attrs) => attrs,
        Err(err) => return err.write_errors(),
    };

    let name = name.unwrap_or_else(|| ident.to_string());

    quote! {
        impl #zod_core::Namespace for #ident {
            const NAME: &'static str = #name;
        }
    }
}

#[cfg(test)]
mod test {
    use crate::test_utils::TokenStreamExt;

    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn default_name() {
        let input = quote! {
            struct MyNs;
        };

        let expected = quote! {
            impl #zod_core::Namespace for MyNs {
                const NAME: &'static str = "MyNs";
            }

        };

        assert_eq!(
            expand(input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        );
    }

    #[test]
    fn custom_name() {
        let input = quote! {
            #[zod(name = "Custom_Name")]
            struct MyNs;
        };

        let expected = quote! {
            impl #zod_core::Namespace for MyNs {
                const NAME: &'static str = "Custom_Name";
            }

        };

        assert_eq!(
            expand(input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        );
    }
}
