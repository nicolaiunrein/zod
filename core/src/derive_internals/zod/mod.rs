mod ast;
mod attrs;
mod custom_suffix;
mod data;
mod r#enum;
mod fields;
mod generics;
mod r#struct;
mod variant;

use self::ast::Ast;
use crate::utils::zod_core;
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use serde_derive_internals::Derive as SerdeDerive;
use syn::DeriveInput;

#[derive(Clone, PartialEq, Eq, Debug, Copy)]
pub enum Derive {
    Input,
    Output,
}

impl From<Derive> for SerdeDerive {
    fn from(value: Derive) -> Self {
        match value {
            Derive::Input => SerdeDerive::Deserialize,
            Derive::Output => SerdeDerive::Serialize,
        }
    }
}

impl ToTokens for Derive {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            Self::Input => tokens.extend(quote!(#zod_core::Kind::Input)),
            Self::Output => tokens.extend(quote!(#zod_core::Kind::Output)),
        }
    }
}

/// convert input into the generated code providing a `Derive`.
pub fn expand(derive: Derive, input: TokenStream) -> TokenStream {
    let derive_input: DeriveInput = match syn::parse2(input) {
        Ok(parsed) => parsed,
        Err(err) => {
            return err.into_compile_error();
        }
    };

    let ast = match Ast::new(derive, derive_input) {
        Ok(attrs) => attrs,
        Err(err) => return err,
    };

    ast.into_token_stream()
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::{
        derive_internals::zod::{
            custom_suffix::CustomSuffix,
            fields::{FieldValue, ZodNamedFieldImpl},
            r#struct::ZodObjectImpl,
        },
        test_utils::TokenStreamExt,
    };
    use pretty_assertions::assert_eq;
    use quote::quote;
    use syn::parse_quote;

    #[test]
    fn expand_ok() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            struct Test {
                inner_string: String,
                inner_u8: u8
            }
        };

        let inner = ZodObjectImpl {
            fields: vec![
                ZodNamedFieldImpl {
                    name: String::from("inner_string"),
                    optional: false,
                    derive,
                    value: FieldValue::Type(parse_quote!(String)),
                },
                ZodNamedFieldImpl {
                    name: String::from("inner_u8"),
                    optional: false,
                    derive,
                    value: FieldValue::Type(parse_quote!(u8)),
                },
            ],
        };

        let custom_suffix = CustomSuffix { inner: None };

        let expected = quote! {
            const _: () = {
                impl #zod_core::Type<#derive> for Test {
                    type Ns = Ns;
                    const NAME: &'static str = "Test";
                    const INLINE: bool = false;

                    fn value() -> #zod_core::z::ZodType<#derive> {
                        #zod_core::z::ZodType {
                            optional: false,
                            custom_suffix: #custom_suffix,
                            inner: #inner.into()
                        }
                    }

                    fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                        let mut v = ::std::vec::Vec::new();
                        v
                    }

                    fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
                }

                impl Ns {
                    #[allow(dead_code)]
                    #[allow(non_upper_case_globals)]
                    const __ZOD_PRIVATE_INPUT___Test: () = {};
                }
            };
        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        )
    }
}
