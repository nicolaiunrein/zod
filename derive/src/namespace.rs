use crate::docs::RustDocs;
use crate::utils::get_zod;
use darling::{FromAttributes, FromDeriveInput, ToTokens};
use quote::quote;

#[derive(FromDeriveInput)]
#[darling(attributes(namespace), forward_attrs(allow, doc, cfg))]
pub(crate) struct Namespace {
    pub vis: syn::Visibility,
    pub ident: syn::Ident,
    pub name: Option<String>,
    pub generics: syn::Generics,
    pub attrs: Vec<syn::Attribute>,
    pub doc: Option<String>,
}

impl ToTokens for Namespace {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let ident = &self.ident;

        let name = &self
            .name
            .as_ref()
            .cloned()
            .unwrap_or_else(|| ident.to_string());

        let docs = RustDocs::from_attributes(&self.attrs).unwrap();

        let zod = get_zod();

        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();

        let expanded = quote! {
            impl #impl_generics #zod::core::Namespace for #ident #ty_generics #where_clause {
                const NAME: &'static str = #name;
                const DOCS: Option<#zod::core::ast::Docs> = #docs;
            }
        };

        tokens.extend(expanded)
    }
}
