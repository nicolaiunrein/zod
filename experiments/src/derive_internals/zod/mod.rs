mod ast;
mod attrs;
mod custom_suffix;
mod r#enum;
mod fields;
mod generics;
mod r#struct;

use self::ast::Ast;
use crate::utils::zod_core;
use proc_macro2::TokenStream as TokenStream2;
use quote::{quote, ToTokens};
use syn::DeriveInput;

#[derive(Clone, PartialEq, Eq, Debug, Copy)]
pub enum Derive {
    Input,
    Output,
}

impl ToTokens for Derive {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        match self {
            Self::Input => tokens.extend(quote!(#zod_core::Kind::Input)),
            Self::Output => tokens.extend(quote!(#zod_core::Kind::Output)),
        }
    }
}

/// convert input into the generated code providing a `Derive`.
pub fn expand(derive: Derive, input: TokenStream2) -> TokenStream2 {
    let derive_input: DeriveInput = match syn::parse2(input) {
        Ok(parsed) => parsed,
        Err(err) => {
            return err.into_compile_error();
        }
    };

    let ast = match Ast::new(derive, derive_input) {
        Ok(attrs) => attrs,
        Err(err) => return err,
    };

    ast.into_token_stream()
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::{
        derive_internals::zod::{
            custom_suffix::CustomSuffix,
            r#enum::EnumImpl,
            r#struct::{ZodObjectImpl, ZodTupleImpl},
        },
        test_utils::TokenStreamExt,
    };
    use pretty_assertions::assert_eq;
    use quote::quote;
    use syn::parse_quote;

    #[test]
    fn expand_zod_for_struct_with_named_fields_ok() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            struct Test {
                inner_string: String,
                inner_u8: u8
            }
        };

        let inner = ZodObjectImpl::new(
            derive,
            &parse_quote!({ inner_string: String, inner_u8: u8 }),
        );

        let custom_suffix = CustomSuffix { inner: None };

        let expected = quote! {
            impl #zod_core::Type<#derive> for Test {
                type Ns = Ns;
                const NAME: &'static str = "Test";
                const INLINE: bool = false;

                fn value() -> #zod_core::z::ZodType<#derive> {
                    #zod_core::z::ZodType {
                        optional: false,
                        custom_suffix: #custom_suffix,
                        inner: #inner.into()
                    }
                }

                fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                    let mut v = ::std::vec::Vec::new();
                    v
                }

                fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
            }

            impl Ns {
                #[allow(dead_code)]
                #[allow(non_upper_case_globals)]
                const __ZOD_PRIVATE_INPUT___Test: () = {};
            }

        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        )
    }

    #[test]
    fn expand_zod_for_struct_with_tuple_fields_ok() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            struct Test(String, u8);
        };

        let custom_suffix = CustomSuffix { inner: None };

        let inner = ZodTupleImpl::new(derive, &parse_quote!((String, u8)));

        let expected = quote! {
            impl #zod_core::Type<#derive> for Test {
                type Ns = Ns;
                const NAME: &'static str = "Test";
                const INLINE: bool = false;

                fn value() -> #zod_core::z::ZodType<#derive> {
                    #zod_core::z::ZodType {
                        optional: false,
                        custom_suffix: #custom_suffix,
                        inner: #inner.into()
                    }
                }

                fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                    let mut v = ::std::vec::Vec::new();
                    v
                }

                fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
            }

            impl Ns {
                #[allow(dead_code)]
                #[allow(non_upper_case_globals)]
                const __ZOD_PRIVATE_INPUT___Test: () = {};
            }

        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        )
    }

    #[test]
    fn impl_zod_for_enum() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            enum Test {
                Unit,
                Tuple1(String),
                Tuple2(String, u8),
                Struct0 {},
                Struct1 {
                    inner: String,
                },
                Struct2 {
                    inner_string: String,
                    inner_u8: u8,
                }
            }
        };

        let inner = EnumImpl::new(
            derive,
            Default::default(),
            vec![
                parse_quote!(Unit),
                parse_quote!(Tuple1(String)),
                parse_quote!(Tuple2(String, u8)),
                parse_quote!(Struct0 {}),
                parse_quote!(Struct1 { inner: String }),
                parse_quote!(Struct2 {
                    inner_string: String,
                    inner_u8: u8
                }),
            ],
        );

        let custom_suffix = CustomSuffix { inner: None };

        let expected = quote! {
            impl #zod_core::Type<#derive> for Test {
                type Ns = Ns;
                const NAME: &'static str = "Test";
                const INLINE: bool = false;

                fn value() -> #zod_core::z::ZodType<#derive> {
                    #zod_core::z::ZodType {
                        optional: false,
                        custom_suffix: #custom_suffix,
                        inner: #inner.into()
                    }
                }

                fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                    let mut v = ::std::vec::Vec::new();
                    v
                }

                fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
            }

            impl Ns {
                #[allow(dead_code)]
                #[allow(non_upper_case_globals)]
                const __ZOD_PRIVATE_INPUT___Test: () = {};
            }

        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        );
    }
}
