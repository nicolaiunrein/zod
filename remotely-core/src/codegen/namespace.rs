use std::collections::{BTreeSet, HashSet};

type RuntimeValue<T> = &'static (dyn Fn() -> T + Sync);

pub trait Namespace {
    const NAME: &'static str;
    type Req: serde::de::DeserializeOwned;

    fn code() -> String {
        let members =
            || inventory::iter::<NsMember>().filter(|member| member.ns_name() == Self::NAME);

        let member_code = members().map(|member| member.decl()).collect::<String>();

        let dependencies: HashSet<String> = members().flat_map(|m| m.deps()).collect();

        let imports = dependencies
            .iter()
            .map(|dep| format!("import {{ {dep} }} from \"./{dep}\";\n"))
            .collect::<String>();

        let export = format!("export namespace {} {{\n{member_code}}};", Self::NAME);

        format!("{imports}{export}")
    }
}
pub enum NsMember {
    Interface {
        ns_name: &'static str,
        name: &'static str,
        raw_decl: RuntimeValue<String>,
        raw_deps: RuntimeValue<Vec<ts_rs::Dependency>>,
    },
    Method {
        ns_name: &'static str,
        name: &'static str,
        args: RuntimeValue<Vec<(&'static str, String)>>,
        res: RuntimeValue<String>,
        raw_deps: RuntimeValue<Vec<ts_rs::Dependency>>,
    },
}

inventory::collect!(NsMember);

impl NsMember {
    pub fn decl(&self) -> String {
        match self {
            NsMember::Interface {
                ns_name,
                name,
                raw_decl,
                ..
            } => {
                let raw = (raw_decl)();
                let full_name = format!("{ns_name}.{name}");
                let decl = raw.replace(&full_name, name);
                format!("export {decl};\n")
            }
            NsMember::Method {
                name, args, res, ..
            } => {
                let args = (args)();
                let res = (res)();

                let args = args
                    .into_iter()
                    .map(|(name, ty)| format!("{name}: {ty}"))
                    .collect::<Vec<_>>()
                    .join(",");

                format!(
                    "export function {name}({args}): Promise<{res}> {{
                    return request(\"Watchout\", \"hello\", arguments);
                }};"
                )
            }
        }
    }

    pub fn deps(&self) -> BTreeSet<String> {
        match self {
            NsMember::Interface { raw_deps, .. } | NsMember::Method { raw_deps, .. } => {
                (raw_deps)()
                    .into_iter()
                    .map(|dep| {
                        dep.ts_name
                            .split_once('.')
                            .expect("dependency has shape `namespace.name`")
                            .0
                            .to_string()
                    })
                    .collect()
            }
        }
    }

    pub fn ns_name(&self) -> &str {
        match self {
            NsMember::Interface { ns_name, .. } => ns_name,
            NsMember::Method { ns_name, .. } => ns_name,
        }
    }
}
