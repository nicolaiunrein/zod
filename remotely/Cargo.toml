[package]
name = "remotely"
version = "0.1.0"
edition = "2021"

[dependencies]
async-trait = "0.1.64"

remotely-core = { path = "../remotely-core" }
serde_json = "1.0.92"
tokio = { version = "1.25.0", features = ["macros", "rt-multi-thread", "time"] }
futures = "0.3.26"
hyper = "0.14.24"
tracing = "0.1.37"
axum = { version = "0.6.7", features = ["ws", "headers"] }
thiserror = "1.0.38"
pin-project-lite = "0.2.9"
