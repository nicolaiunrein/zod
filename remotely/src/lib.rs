pub mod clients;
pub mod servers;

pub use remotely_core::error::Error;
pub use remotely_core::server::Backend;
pub use remotely_core::server::SubscriberMap;
pub use remotely_core::Request;
pub use remotely_core::Response;
pub use remotely_zod::*;

#[async_trait::async_trait]
pub trait Server {
    async fn serve<T>(self, backend: T) -> Result<(), Box<dyn std::error::Error>>
    where
        T: Backend + Send,
        Self: Sized;
}

pub mod __private {
    pub use remotely_core::*;
}

#[test]
fn ui() {
    let t = trybuild::TestCases::new();
    t.compile_fail("tests/ui/fail/*.rs");
    // t.pass("tests/ui/pass/*.rs");
}
