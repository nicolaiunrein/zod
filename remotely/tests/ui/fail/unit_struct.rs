use remotely::zod;
use remotely_core::codegen::namespace::Namespace;

#[derive(zod)]
#[zod(namespace = "Ns")]
struct Test;

struct Ns {}

impl Namespace for Ns {
    const NAME: &'static str = "Ns";
    type Req = NsReq;
}

#[derive(serde::Deserialize)]
struct NsReq;

fn main() {}
