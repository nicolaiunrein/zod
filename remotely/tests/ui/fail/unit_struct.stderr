error: Unsupported shape `no fields`. Expected named fields or one unnamed field.
 --> tests/ui/fail/unit_struct.rs:4:10
  |
4 | #[derive(zod)]
  |          ^^^
  |
  = note: this error originates in the derive macro `zod` (in Nightly builds, run with -Z macro-backtrace for more info)
